File: data/sorting/radix-sort.py
def countingSortForRadix(nums: list[int], exp: int):
    n = len(nums)
    output = [0] * n
    count = [0] * 10

    for i in range(n):
        index = nums[i] // exp
        count[index % 10] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    i = n - 1
    while i >= 0:
        index = nums[i] // exp
        output[count[index % 10] - 1] = nums[i]
        count[index % 10] -= 1
        i -= 1

    for i in range(n):
        nums[i] = output[i]

def radixSort(nums: list[int]) -> list[int]:
    """
    Radix Sort Algorithm
    Time Complexity: O(n * d)
    Space Complexity: O(n + k)
    """
    if not nums:
        return nums
    max_val = max(nums)
    exp = 1
    while max_val // exp > 0:
        countingSortForRadix(nums, exp)
        exp *= 10
    return nums

File: data/sorting/radix-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void countingSortForRadix(vector<int>& nums, int exp) {
    int n = nums.size();
    vector<int> output(n);
    vector<int> count(10, 0);

    for (int i = 0; i < n; ++i) {
        count[(nums[i] / exp) % 10]++;
    }
    for (int i = 1; i < 10; ++i) {
        count[i] += count[i - 1];
    }
    for (int i = n - 1; i >= 0; --i) {
        output[count[(nums[i] / exp) % 10] - 1] = nums[i];
        count[(nums[i] / exp) % 10]--;
    }
    for (int i = 0; i < n; ++i) {
        nums[i] = output[i];
    }
}

vector<int> radixSort(vector<int>& nums) {
    if (nums.empty()) return nums;
    int maxVal = *max_element(nums.begin(), nums.end());
    for (int exp = 1; maxVal / exp > 0; exp *= 10) {
        countingSortForRadix(nums, exp);
    }
    return nums;
}
